### DISCUSSION TITLE
Where can I find a full Shadow topology graphml file?

### DISCUSSION THREAD

**Discussion by ygnkim** [shadow/shadow#673] [Q&A] [answered]
I'm just wondering whether there exists an accessible (realistic) internet topology config file.
In Usenix'14 paper (https://www.robgjansen.com/publications/kist-umntr14-012.pdf),
it is mentioned that they created a realistic Internet map that includes 699,029 vertices and 1,338,590 edges.
I presumed that it must exist in the format of shadow network configuration file, 
but i couldn't find such configuration in shadow repository. 
Is there any way to obtain a realistic internet topology (i.e., graphML configuration) for shadow experiment?

Shadow tutorial mentioned about a network configuration file located at ~/.shadow/share/topology.graphml.xml.
It seems that it is a collapsed version of the complete graph as described in src/tools/topology/readme.
Since i'm not familiar with the python scripts provided at the same directory, 
i'm not clearly understand how it can represent the whole realistic internet topology.
So it would be really nice if i can also access to full version of the realistic internet topology config file.

Thank you.

**Comment by robgjansen** [comment]
(I split this into two issues - see #675 about the mailing list question.)

**Comment by robgjansen** [comment, marked as answer]
There is a new paper at CCS'18 describing a more up-to-date version of the Shadow models. The paper is:

Privacy-Preserving Dynamic Learning of Tor Network Traffic
https://www.robgjansen.com/publications/tmodel-ccs2018.pdf

Most importantly, we made the data from the paper available here:
https://tmodel-ccs2018.github.io/

To answer your question directly, a full topology from that paper, for Shadow v1.13.x, is here:
https://tmodel-ccs2018.github.io/data/shadow/network/atlas.201801.shadow113.graphml.xml.xz

Let me know if this works for you.

(And yes, the `~/.shadow/share/topology.graphml.xml` file is a collapsed version and is less accurate than the full topology that I linked above. Use the full topology if you want to run large and realistic simulations.)

**Comment by robgjansen** [comment]
BTW, if you still want the topology from the Usenix'14 paper, I can make that available as well. The file is 1-2 GiB, so I haven't posted it anywhere yet.

**Comment by ygnkim** [comment]
Oh great! I think the recent model from CCS'18 paper is perfectly suitable for my current purpose.

After looking into the full topology file, I have an another quick question. 

If i want a visualization of the given topology map, what is the easiest way to get the geolocation information for each node?

For example, I found that each node in the configuration have a 'citycode'. (data value with key 'd2' )
I wish it is a standardized way of representing a city, but i have no idea what it indicates.

    <node id="103.202.216.209">
      <data key="d1">103.202.216.209</data>
      <data key="d2">1862258</data>
      <data key="d3">JP</data>
      <data key="d4">9525</data>
      <data key="d5">10449</data>
    </node>

It would be really appreciated if you can give me any advice.

Thank you!

**Comment by ygnkim** [comment]
oh, never mind.

I found that I can easily get the location of the city using their ip address using following website. 
https://tools.keycdn.com/geo 

Thanks!

**Comment by pastly** [comment]
(Hi I'm a coauthor)

The city codes are taken from maxmind's `GeoLite2-City.mmdb`. I seemed to have had `GeoLite2-City_20180102/`.

I downloaded a 28 MB `GeoLite2-City.tar.gz` from maxmind which contains the above dated folder which contains the .mmdb database. I can share this tar if it would be helpful, especially since I don't think maxmind is releasing these databases anymore.

I used maxmind's python library to interact this the .mmdb file, and IIRC it only provides a way to get info given an IP address. I don't know if you can give it a city code and get back information from that.
